src/mongoose_traffic_channel.erl

```diff
@@ -0,0 +1,241 @@
+
+-record(state, {traces = #{},
+                tracing = false,
+                current = <<>>,
+                mappings = #{},
+                start_times = #{}}).
+
+%%--------------------------------------------------------------------
+%% Common callbacks for all cowboy behaviours
+%%--------------------------------------------------------------------
+
+init(Req, Opts) ->
+    Peer = cowboy_req:peer(Req),
+    PeerCert = cowboy_req:cert(Req),
+    ?DEBUG("cowboy init: ~p~n", [{Req, Opts}]),
+    AllModOpts = [{peer, Peer}, {peercert, PeerCert} | Opts],
+    %% upgrade protocol
+    {cowboy_websocket, Req, AllModOpts, #{}}.
+
+terminate(_Reason, _Req, _State) ->
+    ok.
+
+%%--------------------------------------------------------------------
+%% cowboy_http_websocket_handler callbacks
+%%--------------------------------------------------------------------
+
+% Called for every new websocket connection.
+websocket_init(Opts) ->
+    ?DEBUG("websocket_init: ~p~n", [Opts]),
+    gen_server:call(mongoose_traffic, {register, self()}),
+    {ok, #state{}}.
+
+% Called when a text message arrives.
+websocket_handle({text, Msg}, State) ->
+    case handle(jiffy:decode(Msg), State) of
+        {Event, State1} ->
+            {reply, reply(Event), State1};
+        {Event, Payload, State1} ->
+            {reply, reply(Event, Payload), State1}
+    end;
+
+websocket_handle({binary, Msg}, State) ->
+    ?DEBUG("Received binary: ~p", [Msg]),
+    {ok, State};
+
+% With this callback we can handle other kind of
+% messages, like binary.
+websocket_handle(Any, State) ->
+    ?DEBUG("Received non-text: ~p", [Any]),
+    {ok, State}.
+
+% Other messages from the system are handled here.
+websocket_info({message, _Dir, _J, _Stanza}, #state{tracing = false} = State) ->
+    {ok, State};
+websocket_info({message, Dir, {Pid, Jid}, Stanza} = Message, State) ->
+    Spid = pid_to_binary(Pid),
+    Now = now_seconds(),
+    {Traces1, Mappings, IsNewMapping} = record_item(Now,
+                                                    Dir,
+                                                    {Spid, Jid},
+                                                    Stanza,
+                                                    State#state.traces,
+                                                    State#state.mappings),
+    State1 = State#state{traces = Traces1},
+    State2 = State1#state{mappings = Mappings},
+    State3 = maybe_store_start_time(Spid, Now, State2),
+    case maps:size(Traces1) of
+        N when N > ?MAX_ACCOUNTS ->
+            force_stop_tracing(State1);
+        _ ->
+            maybe_send_to_user(Now, IsNewMapping, Message, State3)
+    end;
+websocket_info(stop, State) ->
+    {stop, State};
+websocket_info(Info, State) ->
+    ?DEBUG("unknown info: ~p", [Info]),
+    {ok, State}.
+
+force_stop_tracing(State) ->
+    State1 = State#state{tracing = false},
+    M = reply(<<"error">>, #{<<"reason">> => <<"too_many_accounts">>}),
```

Comment:
About the naming: JID is not necessarily equal to an account in XMPP.

Reply:
Oh, wait, or are these "accounts" the traced events?

Reply:
Technically, an account here is a c2s process, which obviously maps 1to1 to full jid.

Reply:
Yes, c2s maps to a full JID but I've never seen it called "account", which is usually a set of user data in the DB. Unnecessary breach of naming convention.